packages/template-retail-react-app/app/hooks/use-datacloud.js

```diff
@@ -0,0 +1,481 @@
+            this._generateEventDetails('identity', 'Profile'),
+            {
+                isAnonymous: args.isGuest,
+                firstName: args.firstName,
+                lastName: args.lastName
+            }
+        )
+
+        let contactPointEmail = null
+        if (args.email) {
+            contactPointEmail = this._concatenateEvents(
+                baseEvent,
+                this._generateEventDetails('contactPointEmail', 'Profile'),
+                {
+                    email: args.email
+                }
+            )
+        }
+
+        const interaction = {
+            events: [
+                identityProfile,
+                ...(contactPointEmail ? [contactPointEmail] : []),
+                ...catalogObjects
+            ]
+        }
+
+        try {
+            this.sdk.webEventsAppSourceIdPost(interaction)
+        } catch (err) {
+            logger.error('Error sending DataCloud event')
+        }
+    }
+
+    /**
+     * Sends a `catalog-object-impression` event to Data Cloud with
+     * additional search result data.
+     *
+     * This method records a `catalog` event type when a shopper completes a
+     * search, logging an impression of the products displayed in the search
+     * results.
+     *
+     * @param {object} searchParams - The searchParams object
+     * @param {object} searchResults - The searchResults object containing an
+     * array of product impressions
+     * @param {object} args - Additional metadata for the event
+     */
+    async sendViewSearchResults(searchParams, searchResults, args) {
+        const baseEvent = this._constructBaseEvent(args)
+
+        const products = searchResults?.hits?.map((product) =>
+            this._constructDatacloudProduct(product)
+        )
+
+        const catalogObjects = products.map((product) => {
+            return this._concatenateEvents(
+                baseEvent,
+                this._generateEventDetails('catalog', 'Engagement'),
+                this._constructBaseSearchResult(searchParams),
+                {
+                    searchResultId: crypto.randomUUID(),
+                    id: product.id,
+                    type: 'Product',
+                    webStoreId: 'pwa',
+                    interactionName: 'catalog-object-impression'
+                }
+            )
+        })
+
+        const identityProfile = this._concatenateEvents(
+            baseEvent,
+            this._generateEventDetails('identity', 'Profile'),
+            {
+                isAnonymous: args.isGuest,
+                firstName: args.firstName,
+                lastName: args.lastName
+            }
+        )
+
+        let contactPointEmail = null
+        if (args.email) {
+            contactPointEmail = this._concatenateEvents(
+                baseEvent,
+                this._generateEventDetails('contactPointEmail', 'Profile'),
+                {
+                    email: args.email
+                }
+            )
```

Comment:
Same with above comment, we can DRY it by creating a function to return the info since the difference between these methods is the `eventType` and `category`